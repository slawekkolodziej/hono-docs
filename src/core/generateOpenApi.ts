// src/core/generateOpenApi.ts
import fs from "node:fs";
import path from "node:path";
import {
  SyntaxKind,
  TypeLiteralNode,
  ImportTypeNode,
  TypeReferenceNode,
  TypeNode,
  ts,
} from "ts-morph";
import type {
  AppTypeSnapshotPath,
  GenerateParams,
  OpenApiPath,
  ApiGroup,
} from "../types";
import { genParameters } from "../utils/parameters";
import { genRequestBody } from "../utils/requestBody";
import { buildSchema } from "../utils/buildSchema";
import { groupBy, unwrapUnion } from "../utils/format";
import { extractDocumentationFromMiddleware, createDocLookup } from "../utils/middleware-docs";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type OpenAPI = Record<string, any>;

/**
 * CORE ALGORITHM: Extract TypeLiteral nodes with path prefixes from Hono's complex type structures
 * 
 * This is the heart of route extraction for grouped Hono apps. Hono's `.route()` method creates
 * complex nested type unions that represent grouped routes with path prefixes.
 * 
 * For example:
 * ```typescript
 * const app = new Hono()
 *   .route("/api", apiRoutes)    // Creates MergeSchemaPath<ApiSchema, "/api">
 *   .route("/auth", authRoutes)  // Creates MergeSchemaPath<AuthSchema, "/auth">
 * ```
 * 
 * This results in union types like:
 * BlankSchema | MergeSchemaPath<ApiSchema, "/api"> | MergeSchemaPath<AuthSchema, "/auth">
 * 
 * We need to:
 * 1. Skip BlankSchema (empty base schema)
 * 2. Extract schemas from MergeSchemaPath and combine them with their path prefixes
 * 3. Handle nested unions/intersections recursively
 * 4. Return all TypeLiteral nodes (route definitions) with their full path prefixes
 */
function extractTypeLiteralsFromUnion(unionNode: TypeNode, currentPrefix: string = ""): Array<{ literal: TypeLiteralNode; prefix: string }> {
  const literals: Array<{ literal: TypeLiteralNode; prefix: string }> = [];

  if (!unionNode.isKind(SyntaxKind.UnionType)) {
    return literals;
  }

  for (const member of unionNode.asKind(SyntaxKind.UnionType)!.getTypeNodes()) {
    console.log("üîç Processing union member:", member.getKindName(), "with prefix:", currentPrefix);

    if (member.isKind(SyntaxKind.TypeLiteral)) {
      console.log("üîç Found direct TypeLiteral in union");
      literals.push({ literal: member as TypeLiteralNode, prefix: currentPrefix });
    } else if (member.isKind(SyntaxKind.ImportType)) {
      // Handle ImportType with qualifier (like import("...").MergeSchemaPath)
      const importType = member.asKind(SyntaxKind.ImportType)!;
      const qualifier = importType.getQualifier();
      console.log("üîç ImportType qualifier:", qualifier?.getKindName());

      if (qualifier) {
        let qualifierName = "";
        if (qualifier.isKind(SyntaxKind.QualifiedName)) {
          const qualifiedName = qualifier.asKind(SyntaxKind.QualifiedName)!;
          qualifierName = qualifiedName.getRight().getText();
        } else if (qualifier.isKind(SyntaxKind.Identifier)) {
          qualifierName = qualifier.asKind(SyntaxKind.Identifier)!.getText();
        }
        console.log("üîç ImportType qualifier name:", qualifierName);

        // Only process MergeSchemaPath, skip BlankSchema
        if (qualifierName === "MergeSchemaPath") {
          const typeArgs = importType.getTypeArguments();
          console.log("üîç MergeSchemaPath typeArgs:", typeArgs.length);
          if (typeArgs.length >= 2) {
            const schema = typeArgs[0];
            const pathPrefix = typeArgs[1].getText().replace(/"/g, "");
            const fullPrefix = currentPrefix + pathPrefix;
            console.log("üîç Schema kind:", schema.getKindName(), "with prefix:", fullPrefix);

            if (schema.isKind(SyntaxKind.IntersectionType)) {
              console.log("üîç Processing IntersectionType with", schema.asKind(SyntaxKind.IntersectionType)!.getTypeNodes().length, "members");
              for (const intersectMember of schema.asKind(SyntaxKind.IntersectionType)!.getTypeNodes()) {
                console.log("üîç Intersection member kind:", intersectMember.getKindName());
                if (intersectMember.isKind(SyntaxKind.TypeLiteral)) {
                  console.log("üîç Found TypeLiteral in IntersectionType!");
                  literals.push({ literal: intersectMember as TypeLiteralNode, prefix: fullPrefix });
                } else {
                  console.log("üîç Non-TypeLiteral in intersection:", intersectMember.getKindName());
                }
              }
            } else if (schema.isKind(SyntaxKind.UnionType)) {
              // Handle nested unions (recursive case)
              const nestedLiterals = extractTypeLiteralsFromUnion(schema, fullPrefix);
              literals.push(...nestedLiterals);
            } else {
              // Recursively extract from the schema
              const nestedLiterals = extractTypeLiteralsFromUnion(schema, fullPrefix);
              console.log("üîç Found", nestedLiterals.length, "nested literals");
              literals.push(...nestedLiterals);
            }
          }
        } else {
          console.log("üîç Skipping non-MergeSchemaPath:", qualifierName);
        }
        // BlankSchema is intentionally skipped (empty schema)
      }
    } else if (member.isKind(SyntaxKind.IntersectionType)) {
      console.log("üîç Processing IntersectionType in union");
      // Handle intersections within unions
      for (const intersectMember of member.asKind(SyntaxKind.IntersectionType)!.getTypeNodes()) {
        console.log("üîç Intersection member:", intersectMember.getKindName());
        if (intersectMember.isKind(SyntaxKind.TypeLiteral)) {
          console.log("üîç Found TypeLiteral in intersection");
          literals.push({ literal: intersectMember as TypeLiteralNode, prefix: currentPrefix });
        }
      }
    } else {
      console.log("üîç Ignoring union member type:", member.getKindName());
    }
  }

  return literals;
}

/**
 * MAIN FUNCTION: Generate OpenAPI specification from Hono AppType
 * 
 * This function orchestrates the entire OpenAPI generation process:
 * 1. Load and parse TypeScript source files
 * 2. Extract documentation from doc() middleware (if present)
 * 3. Parse the Hono AppType to extract route type information
 * 4. Convert TypeScript route types to OpenAPI paths/operations
 * 5. Generate response schemas and request/response examples
 * 6. Write the final OpenAPI JSON specification to disk
 * 
 * @param config - Hono-docs configuration with OpenAPI metadata
 * @param snapshotPath - Path information for the generated AppType snapshot
 * @param apiGroup - API group configuration with route patterns
 * @param fileName - Output filename for the OpenAPI JSON
 * @param project - ts-morph Project instance for TypeScript analysis
 * @param rootPath - Root directory path for resolving relative paths  
 * @param outputRoot - Output directory for generated files
 * @returns Promise resolving to the output path information
 */
export async function generateOpenApi({
  config,
  snapshotPath,
  apiGroup,
  fileName,
  project,
  rootPath,
  outputRoot,
}: GenerateParams & {
  snapshotPath: AppTypeSnapshotPath;
  apiGroup: ApiGroup;
}): Promise<OpenApiPath> {
  const sf = project.addSourceFileAtPath(
    path.resolve(rootPath, snapshotPath.appTypePath)
  );
  
  // Extract documentation from doc() middleware in the original source file
  console.log("üîç Extracting documentation from middleware in source file:", apiGroup.appTypePath);
  let docLookup = new Map<string, import("../utils/middleware-docs").DocConfig>();
  
  try {
    const originalSf = project.addSourceFileAtPath(
      path.resolve(rootPath, apiGroup.appTypePath)
    );
    const documentation = extractDocumentationFromMiddleware(originalSf);
    docLookup = createDocLookup(documentation);
    console.log(`üîç Found ${documentation.length} routes with doc() middleware`);
  } catch (error) {
    console.warn(`‚ö†Ô∏è Failed to extract documentation from middleware: ${error}`);
    console.warn(`‚ö†Ô∏è Continuing with auto-generated documentation only`);
  }
  
  const aliasDecl = sf.getTypeAliasOrThrow("AppType");

  const topTypeNode = aliasDecl.getTypeNode();

  let typeArgs: readonly TypeNode<ts.TypeNode>[];

  if (topTypeNode?.isKind(SyntaxKind.TypeReference)) {
    typeArgs = (topTypeNode as TypeReferenceNode).getTypeArguments();
  } else if (topTypeNode?.isKind(SyntaxKind.ImportType)) {
    typeArgs = (topTypeNode as ImportTypeNode).getTypeArguments();
  } else {
    throw new Error("AppType must be an ImportType or a TypeReference");
  }

  if (typeArgs.length < 2) {
    throw new Error("Expected two type arguments on HonoBase");
  }

  const routesNode = typeArgs[1];

  // Gather all TypeLiteralNodes with their prefixes (handle intersections)
  const literalsWithPrefixes: Array<{ literal: TypeLiteralNode; prefix: string }> = [];
  if (routesNode.isKind(SyntaxKind.IntersectionType)) {
    for (const tn of routesNode
      .asKind(SyntaxKind.IntersectionType)!
      .getTypeNodes()) {
      if (tn.isKind(SyntaxKind.TypeLiteral))
        literalsWithPrefixes.push({ literal: tn as TypeLiteralNode, prefix: "" });
    }
  } else if (routesNode.isKind(SyntaxKind.UnionType)) {
    // Handle unions (for grouped AppType)
    console.log("üîç Processing UnionType with", routesNode.asKind(SyntaxKind.UnionType)!.getTypeNodes().length, "members");

    const unionLiterals = extractTypeLiteralsFromUnion(routesNode);
    literalsWithPrefixes.push(...unionLiterals);
    console.log("üîç Extracted", unionLiterals.length, "TypeLiteral nodes from union");
  } else if (routesNode.isKind(SyntaxKind.TypeLiteral)) {
    literalsWithPrefixes.push({ literal: routesNode as TypeLiteralNode, prefix: "" });
  } else {
    console.log("üîç Routes node kind:", routesNode.getKindName());
    console.log("üîç Routes node text:", routesNode.getText());
    throw new Error("Routes type is not a literal, intersection, or union of literals");
  }

  const paths: OpenAPI = {};

  for (const { literal: lit, prefix } of literalsWithPrefixes) {
    console.log("üîç Processing literal with prefix:", prefix);
    for (const member of lit.getMembers()) {
      if (!member.isKind(SyntaxKind.PropertySignature)) continue;
      const routeProp = member.asKindOrThrow(SyntaxKind.PropertySignature);
      // Extract route string and normalize to OpenAPI path syntax
      const raw = routeProp.getNameNode().getText().replace(/"/g, "");
      // Normalize path concatenation to avoid double slashes
      const routeWithPrefix = (prefix === "/" || prefix === "") 
        ? raw 
        : prefix + raw;
      const route = routeWithPrefix.replace(/:([^/]+)/g, "{$1}");
      console.log("üîç Processing route:", raw, "->", routeWithPrefix, "->", route);
      if (!paths[route]) paths[route] = {};

      // === NEW: get the RHS TypeLiteralNode properly ===
      const tn = routeProp.getTypeNode();
      if (!tn || !tn.isKind(SyntaxKind.TypeLiteral)) continue;
      const rhs = tn as TypeLiteralNode;

      for (const m of rhs.getMembers()) {
        if (!m.isKind(SyntaxKind.PropertySignature)) continue;
        const methodProp = m.asKindOrThrow(SyntaxKind.PropertySignature);
        const name = methodProp.getNameNode().getText(); // e.g. "$get"
        const http = name.slice(1).toLowerCase(); // "get", "post", etc.
        const variants = unwrapUnion(methodProp.getType());

        // Check for doc() middleware documentation for this route and method
        const docKey = `${http}:${raw}`; // Use the original route, not the prefixed one
        const docConfig = docLookup.get(docKey);
        
        console.log(`üîç Looking for doc() middleware with key: ${docKey}`);
        if (docConfig) {
          console.log(`üîç Found doc() middleware for ${http.toUpperCase()} ${raw}:`, docConfig);
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const op: any = {
          summary: docConfig?.summary || `Auto-generated ${http.toUpperCase()} ${route}`,
        };
        
        // Add description if available from doc() middleware
        if (docConfig?.description) {
          op.description = docConfig.description;
        }
        
        // Add tags if available from doc() middleware
        if (docConfig?.tags && docConfig.tags.length > 0) {
          op.tags = docConfig.tags;
        }
        
        // Mark as deprecated if specified in doc() middleware
        if (docConfig?.deprecated) {
          op.deprecated = true;
        }

        // parameters
        const params = genParameters(variants[0]);
        if (params.length) op.parameters = params;

        // requestBody
        const rb = genRequestBody(variants[0]);
        if (rb) op.requestBody = rb;

        // responses
        op.responses = {};
        const byStatus = groupBy(variants, (v) => {
          const s = v
            .getProperty("status")!
            .getValueDeclarationOrThrow()
            .getType()
            .getText();
          return /^\d+$/.test(s) ? s : "default";
        });
        for (const [code, vs] of Object.entries(byStatus)) {
          const schemas = vs.map((v) =>
            buildSchema(
              v.getProperty("output")!.getValueDeclarationOrThrow().getType()
            )
          );
          const schema = schemas.length > 1 ? { oneOf: schemas } : schemas[0];
          op.responses[code] = {
            description:
              code === "default"
                ? `Default fallback response`
                : `Status ${code}`,
            content: { "application/json": { schema } },
          };
        }



        paths[route][http] = op;
      }
    }
  }

  const spec = {
    ...config.openApi,
    paths,
  };

  // write to disk
  const outputPath = path.join(outputRoot, `${fileName}.json`);

  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  fs.writeFileSync(outputPath, JSON.stringify(spec, null, 2), "utf-8");
  console.log(`‚úÖ OpenAPI written to ${outputPath}`);
  return { openApiPath: outputPath };
}