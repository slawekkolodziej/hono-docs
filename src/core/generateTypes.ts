import fs from "node:fs";
import path from "node:path";
import createDebug from "debug";
import type { ApiGroup, GenerateParams } from "../types";
import { normalizeImportPaths } from "../utils/format";

const debug = createDebug("hono-docs");

export async function generateTypes({
  config,
  project,
  rootPath,
  apiGroup,
  fileName,
  outputRoot,
}: GenerateParams & { apiGroup: ApiGroup }) {
  fs.mkdirSync(outputRoot, { recursive: true });

  const outputPath = path.join(outputRoot, `${fileName}.d.ts`);
  const absInput = path.resolve(rootPath, apiGroup.appTypePath);

  const sourceFile = project.addSourceFileAtPath(absInput);
  const typeAliases = sourceFile.getTypeAliases();
  const interfaces = sourceFile.getInterfaces();

  let result = `// AUTO-GENERATED from ${apiGroup.appTypePath}\n\n`;

  typeAliases.forEach((alias) => {
    const raw = alias.getType().getText(alias);
    const clean = normalizeImportPaths(raw);
    result += `export type ${alias.getName()} = ${clean};\n\n`;
  });

  interfaces.forEach((intf) => {
    result += intf.getText() + "\n\n";
  });

  const preContent = config.preDefineTypeContent || "";

  fs.writeFileSync(outputPath, `${preContent}\n${result}`, "utf-8");
  debug("Generated types for %s -> %s", apiGroup.appTypePath, outputPath);
  return { appTypePath: outputPath, name: fileName };
}
