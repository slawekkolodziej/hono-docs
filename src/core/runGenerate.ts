import fs from "node:fs";
import path, { resolve } from "node:path";
import createDebug from "debug";
import { Project } from "ts-morph";
import { loadConfig } from "../config/loadConfig";
import { generateTypes } from "./generateTypes";
import { generateOpenApi } from "./generateOpenApi";
import { Api } from "../types";
import { cleanDefaultResponse, sanitizeApiPrefix } from "../utils/format";
import { getLibDir } from "../utils/libDir";

const debug = createDebug("hono-docs");

export async function runGenerate(configPath: string) {
  const config = await loadConfig(configPath);

  const rootPath = process.cwd();
  debug("Initializing ts-morph with tsConfig: %s", config.tsConfigPath);
  const project = new Project({
    tsConfigFilePath: resolve(rootPath, config.tsConfigPath),
  });

  // const isDevMode =
  //   __dirname.includes("/src/") || __dirname.includes("\\src\\");

  // const libDir = isDevMode
  //   ? path.resolve(__dirname, "../../")
  //   : // : path.dirname(require.resolve("@rcmade/hono-docs/package.json"));
  //     path.dirname(fileURLToPath(import.meta.url));
  const libDir = getLibDir();
  debug("Library root directory: %s", libDir);

  const apis = config.apis;

  const snapshotOutputRoot = path.resolve(libDir, "output/types");
  const openAPiOutputRoot = path.resolve(libDir, "output/openapi");

  const commonParams = {
    config,
    libDir,
    project,
    rootPath,
  };
  for (const apiGroup of apis) {
    const sanitizedName = sanitizeApiPrefix(apiGroup.apiPrefix);

    const snapshotPath = await generateTypes({
      ...commonParams,
      apiGroup: apiGroup,
      fileName: sanitizedName,
      outputRoot: snapshotOutputRoot,
    });

     debug("About to call generateOpenApi for: %s (source: %s)", snapshotPath.appTypePath, apiGroup.appTypePath);
     await generateOpenApi({
       snapshotPath,
       apiGroup,
       ...commonParams,
       fileName: sanitizedName,
       outputRoot: openAPiOutputRoot,
     });
     debug("generateOpenApi completed for: %s", apiGroup.appTypePath);
  }

  const merged = {
    ...config.openApi,
    tags: [] as { name: string }[],
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    paths: {} as Record<string, any>,
  };

  for (const apiGroup of apis) {
    const name = sanitizeApiPrefix(apiGroup.apiPrefix);
    const openApiFile = path.join(openAPiOutputRoot, `${name}.json`);

    if (!fs.existsSync(openApiFile)) {
      console.warn(`⚠️ Missing OpenAPI file: ${openApiFile}`);
      continue;
    }

    const json = JSON.parse(fs.readFileSync(openApiFile, "utf-8"));
    merged.tags.push({ name: apiGroup.name });

    const customApiMap = new Map<string, Api>();

    if (apiGroup?.api) {
      for (const customApi of apiGroup.api) {
        const fullPath =
          path.posix
            .join(apiGroup.apiPrefix, customApi.api)
            .replace(/\/+$/, "") || "/";
        customApiMap.set(
          `${customApi.method.toLowerCase()} ${fullPath}`,
          customApi
        );
      }
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    for (const [pathKey, operations] of Object.entries<any>(json.paths)) {
      const prefixedPath =
        path.posix.join(apiGroup.apiPrefix, pathKey).replace(/\/+$/, "") || "/";
      if (!merged.paths[prefixedPath]) merged.paths[prefixedPath] = {};

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      for (const [method, operation] of Object.entries<any>(operations)) {
        const opKey = `${method.toLowerCase()} ${prefixedPath}`;
        const customApi = customApiMap.get(opKey);

        // Override or enrich metadata if defined
        if (customApi) {
          operation.summary = customApi.summary || operation.summary;
          operation.description =
            customApi.description || operation.description;
          operation.tags =
            customApi.tag && customApi.tag.length > 0
              ? customApi.tag
              : [apiGroup.name];
        } else {
          operation.tags = operation.tags || [];
          if (!operation.tags.includes(apiGroup.name)) {
            operation.tags.push(apiGroup.name);
          }
        }

        cleanDefaultResponse(operation, prefixedPath, method);
        merged.paths[prefixedPath][method] = operation;
      }
    }
  }

  const outputPath = path.join(rootPath, config.outputs.openApiJson);
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });

  fs.writeFileSync(outputPath, `${JSON.stringify(merged, null, 2)}\n`);

  console.log(`✅ OpenAPI spec written to: ${outputPath}`);
}
