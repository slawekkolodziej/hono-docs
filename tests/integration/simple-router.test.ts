import { describe, test, expect, beforeAll } from "vitest";
import fs from "fs";
import path from "path";
import { runGenerate } from "../../src/core/runGenerate";

// Test files are located in integration-tests directory
const rootDir = path.resolve(import.meta.dirname, "../..");
const integrationTestsDir = path.join(rootDir, "integration-tests");

describe("manual-routes - Manual API definitions with overrides", () => {
  const configPath = "./manual-routes/hono-docs.ts";
  const configPathFull = path.join(integrationTestsDir, configPath);
  const testCaseDir = path.dirname(configPathFull);
  const outputPath = path.resolve(testCaseDir, "openapi.json");
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let output: any = null;

  beforeAll(async () => {
    // Change to the test case directory to match CLI behavior
    const originalCwd = process.cwd();
    process.chdir(testCaseDir);
    
    try {
      await runGenerate(configPathFull);
      output = JSON.parse(fs.readFileSync(outputPath, "utf-8"));
    } finally {
      // Always restore original working directory
      process.chdir(originalCwd);
    }
  });

  test("generate OpenAPI spec", () => {
    // Check that output file exists
    expect(fs.existsSync(outputPath)).toBe(true);

    // Basic structure checks
    expect(output).toHaveProperty("openapi");
    expect(output).toHaveProperty("info");
    expect(output).toHaveProperty("paths");

    // Check expected paths exist
    expect(output.paths).toHaveProperty("/user");
    expect(output.paths).toHaveProperty("/user/{id}");

    // Check manual tag overrides work
    expect(output.paths["/user"].get.tags).toEqual(["UserList"]);
    expect(output.paths["/user/{id}"].get.tags).toEqual(["User Routes"]);

    // Check parameter extraction
    expect(output.paths["/user/{id}"].get.parameters).toHaveLength(1);
    expect(output.paths["/user/{id}"].get.parameters[0].name).toBe("id");
    expect(output.paths["/user/{id}"].get.parameters[0].in).toBe("path");
  });

  test("manual configuration overrides auto-generation", () => {
    // UserList tag from manual config should override default
    expect(output.paths["/user"].get.tags).toEqual(["UserList"]);
    
    // Second route should get the apiGroup name as fallback
    expect(output.paths["/user/{id}"].get.tags).toEqual(["User Routes"]);
  });

  test("matches snapshot", () => {
    expect(output).toMatchSnapshot();
  });
});

describe("simple-router - Non-grouped basic Hono app", () => {
  const configPath = "./simple-router/hono-docs.config.ts";
  const configPathFull = path.join(integrationTestsDir, configPath);
  const testCaseDir = path.dirname(configPathFull);
  const outputPath = path.resolve(testCaseDir, "openapi.json");

  test("generates complete OpenAPI spec with simple router support", async () => {
    // FIXED: Simple Hono apps now work! 
    // 
    // When TypeScript generates BlankSchema for simple apps, we fall back to 
    // source code parsing to extract routes directly from the AST.
    //
    // This generates complete OpenAPI with all routes and HTTP methods.
    
    // Change to the test case directory to match CLI behavior
    const originalCwd = process.cwd();
    process.chdir(testCaseDir);
    
    try {
      await runGenerate(configPathFull);
    } finally {
      // Always restore original working directory
      process.chdir(originalCwd);
    }

    expect(fs.existsSync(outputPath)).toBe(true);
    
    const output = JSON.parse(fs.readFileSync(outputPath, "utf-8"));
    expect(output).toHaveProperty("openapi");
    expect(output).toHaveProperty("info");
    expect(output).toHaveProperty("paths");
    
    // Now generates complete paths from source parsing
    expect(Object.keys(output.paths)).toHaveLength(2); // /api/users and /api/users/{id}
    
    // Verify the generated routes match our simple router implementation
    expect(output.paths).toHaveProperty("/api/users");
    expect(output.paths).toHaveProperty("/api/users/{id}");
    
    // Verify HTTP methods are detected
    expect(output.paths["/api/users"]).toHaveProperty("get");
    expect(output.paths["/api/users"]).toHaveProperty("post"); 
    expect(output.paths["/api/users/{id}"]).toHaveProperty("get");
    expect(output.paths["/api/users/{id}"]).toHaveProperty("put");
    expect(output.paths["/api/users/{id}"]).toHaveProperty("delete");
    
    // Verify proper tagging from API group
    expect(output.paths["/api/users"].get.tags).toEqual(["Simple Router"]);
    expect(output.paths["/api/users/{id}"].get.tags).toEqual(["Simple Router"]);
  });

  test("validates complete OpenAPI structure for simple router", async () => {
    // This test validates the complete OpenAPI structure and content
    // for simple Hono apps that don't use .route() grouping
    
    // Change to the test case directory to match CLI behavior
    const originalCwd = process.cwd();
    process.chdir(testCaseDir);
    
    try {
      await runGenerate(configPathFull);
    } finally {
      // Always restore original working directory
      process.chdir(originalCwd);
    }

    const output = JSON.parse(fs.readFileSync(outputPath, "utf-8"));

    // Basic structure validation
    expect(output).toHaveProperty("openapi", "3.0.0");
    expect(output).toHaveProperty("info");
    expect(output.info).toEqual({
      title: "Simple Router Test API",
      version: "1.0.0",
      description: "Simple non-grouped router pattern test"
    });

    // Tags should be generated from API group name
    expect(output).toHaveProperty("tags");
    expect(output.tags).toContainEqual({
      name: "Simple Router"
    });

    // All routes from routes.ts should be present
    expect(output.paths).toHaveProperty("/api/users");
    expect(output.paths).toHaveProperty("/api/users/{id}");

    // Validate GET /api/users route
    const getUsersRoute = output.paths["/api/users"].get;
    expect(getUsersRoute).toBeDefined();
    expect(getUsersRoute.tags).toEqual(["Simple Router"]);
    expect(getUsersRoute.responses).toHaveProperty("default");
    expect(getUsersRoute.responses["default"]).toHaveProperty("content");

    // Validate GET /api/users/{id} route
    const getUserByIdRoute = output.paths["/api/users/{id}"].get;
    expect(getUserByIdRoute).toBeDefined();
    expect(getUserByIdRoute.tags).toEqual(["Simple Router"]);
    // Note: Simple router fallback doesn't generate parameter schemas from source parsing

    // Validate POST /api/users route
    const postUsersRoute = output.paths["/api/users"].post;
    expect(postUsersRoute).toBeDefined();
    expect(postUsersRoute.tags).toEqual(["Simple Router"]);
    expect(postUsersRoute.responses).toHaveProperty("201");
    // Note: With proper method chaining, TypeScript can infer response types and status codes

    // Validate PUT /api/users/{id} route
    const putUserByIdRoute = output.paths["/api/users/{id}"].put;
    expect(putUserByIdRoute).toBeDefined();
    expect(putUserByIdRoute.tags).toEqual(["Simple Router"]);

    // Validate DELETE /api/users/{id} route
    const deleteUserByIdRoute = output.paths["/api/users/{id}"].delete;
    expect(deleteUserByIdRoute).toBeDefined();
    expect(deleteUserByIdRoute.tags).toEqual(["Simple Router"]);
  });

  test("matches complete simple router structure snapshot", () => {
    const output = JSON.parse(fs.readFileSync(outputPath, "utf-8"));
    expect(output).toMatchSnapshot();
  });
});