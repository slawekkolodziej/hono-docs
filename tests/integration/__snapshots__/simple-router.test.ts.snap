// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`manual-routes - Manual API definitions with overrides > matches snapshot 1`] = `
{
  "info": {
    "description": "Manual route definition pattern test",
    "title": "Manual Routes Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/user": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "message",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /user",
        "tags": [
          "UserList",
        ],
      },
    },
    "/user/{id}": {
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "userId": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "userId",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /user/{id}",
        "tags": [
          "User Routes",
        ],
      },
    },
  },
  "servers": [
    {
      "url": "http://localhost:3000/api",
    },
  ],
  "tags": [
    {
      "name": "User Routes",
    },
  ],
}
`;

exports[`simple-router - Non-grouped basic Hono app > matches complete simple router structure snapshot 1`] = `
{
  "info": {
    "description": "Simple non-grouped router pattern test",
    "title": "Simple Router Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/api/users": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "users": {
                      "items": {},
                      "type": "array",
                    },
                  },
                  "required": [
                    "users",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /api/users",
        "tags": [
          "Simple Router",
        ],
      },
      "post": {
        "responses": {
          "201": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "created": {},
                  },
                  "required": [
                    "created",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Status 201",
          },
        },
        "summary": "Auto-generated POST /api/users",
        "tags": [
          "Simple Router",
        ],
      },
    },
    "/api/users/{id}": {
      "delete": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "deleted": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "deleted",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated DELETE /api/users/{id}",
        "tags": [
          "Simple Router",
        ],
      },
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "user": {
                      "properties": {
                        "id": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "id",
                      ],
                      "type": "object",
                    },
                  },
                  "required": [
                    "user",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /api/users/{id}",
        "tags": [
          "Simple Router",
        ],
      },
      "put": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "updated": {},
                  },
                  "required": [
                    "updated",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated PUT /api/users/{id}",
        "tags": [
          "Simple Router",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "Simple Router",
    },
  ],
}
`;
